src/02-common-principles.md

```diff
@@ -367,6 +367,102 @@ In particular, if a BIDS dataset contains a `derivatives/` sub-directory,
 the contents of that directory may be a heterogeneous mix of BIDS Derivatives
 datasets and non-compliant derivatives.
 
+## BIDS URI: Pointing to files within and outside of BIDS datasets
+
+Throughout the BIDS specification there a meta data fields that contain a
+pointer to a file.
+For example `IntendedFor`, `SpatialReference`, and `AssociatedEmptyRoom`.
+Prior to BIDS version 1.6.1, such pointers were required to be "relative paths".
+However the specification failed to consistently  describe relative to *what*
+these pointers should be specified, and the exact syntax to be used.
+Yet even with perfectly well and consistently defined "relative paths", this
+method of pointing to files becomes impractical when referring to files that
+are outside of a particular file tree (for example, on the Internet), or when
+the location of the data that is being pointed to changes.
+
+For these reasons, starting with BIDS version 1.6.1, using "relative paths" for
+all fields pointing to a separate file is [DEPRECATED][].
+
+Instead, the following [URI][uniform-resource-indicator] scheme SHOULD be
+used to point to files within and outside of BIDS datasets, as elaborated below.
+
+`bids:<dataset-name>:/absolute/path/within/dataset`
+
+Here, `bids:` is the URI scheme, and `<dataset-name>:/absolute/path/within/dataset`
+is the path to the file, consisting of the name of a dataset, a colon, and the absolute
+path within that dataset that MUST start with a forward slash `/`.
+
+The location of a dataset with the name `<dataset-name>` MUST be specified in the
+[`dataset_description.json` file](./03-modality-agnostic-files.md#dataset-description)
+under the `DatasetLinks` field, which is an [object][] of [strings][], as shown in
+the examples below.
+Note that the `<dataset-name>`: `"local"` is a reserved value that may only be
+used to refer to the current dataset.
+
+### Refer to a file within a dataset
+
+To link for example to a derivative file that is stored in a nested `derivatives/`
+directory within a dataset:
+`bids:deriv1:/sub-01/anat/sub-01_desc-preproc_T1w.nii.gz`
```

Comment:
`bids:deriv1:::/sub-01/anat/sub-01_desc-preproc_T1w.nii.gz`

just a small suggestion to allow for a few extra `:` inspired by (https://docs.aws.amazon.com/general/latest/gr/aws-arns-and-namespaces.html)

potentially allow for external unique dataset references without having to create a json file.

```
bids:openneuro:000003:2.0.1:/sub-01/anat/sub-01_desc-preproc_T1w.nii.gz
bids:dandi:000056:draft:/sub-01/anat/sub-01_desc-preproc_T1w.nii.gz
```

Reply:
Thanks for the suggestion, I see the convenience of not having to specify a `dataset_description.DatasetLinks` object.

However I still think it's better to REQUIRE that object for anything but `local` --> because if the location of some data `deriv1` changes, then all you need to do is update a single field in `dataset_description.DatasetLinks` as opposed to doing a filetree-wide search and replace.

Reply:
@sappelhoff - this suggestion was specifically about external datasets that do not change. i.e. one should never have to update unless you are changing the reference dataset itself, which presumably is less common. also this will effectively in the future allow for an URI scheme.

i'm in general fine with prefixes of all kinds whose metadata are elsewhere. this was mostly a suggestion that allows compressed forms, which is effectively what the current proposal is, with a more schemafied form like a uri scheme for known entities (kind of analogous to the space references in bids).

now i can see the potential for operations on such things, where a person says, i want an external dataset, but i have it locally, so i will just change the description to point to a local path.  and technically that should still be possible, but that can be an implementer's responsibility rather than the specification. 

so perhaps look at it as the `:` allowing a future scheme, even if the current scheme is implemented as proposed, so you have future compatibility in place rather than preventing it because it would break backwards compatibility. for present, you may simply say the `:`  are required and that the specification reserves the possibility of extending this to support known dataset locations.

Reply:
Thanks for this explanation Satra, I think your argument to allow for future compatibility makes a lot of sense, so I at least agree that we should adjust the present proposal to allow for a flexible, backward-compatible extension in the future.

I am still a bit confused about this part:

> you may simply say the `:` are required and that the specification reserves the possibility of extending this to support known dataset locations.

In your suggestion you put 2 "extra" colons after `<dataset-name>`: `bids:deriv1:::/sub-01/anat/sub-01_desc-preproc_T1w.nii.gz`

However from your linked aws resource (https://docs.aws.amazon.com/general/latest/gr/aws-arns-and-namespaces.html) and the Wikipedia page on URIs (https://en.wikipedia.org/wiki/Uniform_Resource_Identifier) it's not clear to me:

1. how may "extra colons" we should reserve for now
1. how the space between the colons shall be used in the future, and to what extent this needs to be defined now

Reply:
since this is proposing a schema, this is completely up to us: here the uriparser would say:

scheme = bids
path = rest

(an example is `mailto:user@example.org` / https://en.wikipedia.org/wiki/Mailto)

and the only things that would interpret the path would be bids tools. the `:` is just one example of a path delineation especially given that `/` may be part of the filepath.

so a bids schema could look like:

`bids:<namespace | host>:<datasetid>:<version>:/path`

i haven't thought this through fully. but it's up to us to decide what the bids scheme should look like and we could cover local references both within and outside a given dataset, as well as absolutely references like external datasets/services.

here is a list of uri schemes that we could take some inspiration from: https://en.wikipedia.org/wiki/List_of_URI_schemes

Reply:
The problem with any URI scheme is resolution, and who has to do it. Stefan's and my proposal basically splits the problem into `(name, relpath)`, and then uses a `name -> URI` mapping to punt on resolution entirely. People can put in a `git:` URI to a datalad dataset, an `https:` URI to a landing page in a database, or whatever pleases them. If I need to resolve it, BIDS has no need to tell me how.

As soon as we start having hosts, then we need a way to translate a `(host, datasetid)` into something that can be retrieved. It might be a simple translation, but if it's not building on an existing standard, then we either need to specify a translation or maintain a table of translations that tools can implement. Thus I preferred to avoid this.

Now, if we're going to add hosts, there is a notion of "authority" built into URI.

```
URI = scheme:[//authority]path[?query][#fragment]
authority = [userinfo@]host[:port]
```

So we'd get:

```
bids:[//[userinfo@]host[:port]]dataset:relativepath
```


The default authority would be to search `DatasetLinks` in the local dataset. If an authority is provided, it would have to be the responsibility of that authority to provide a path specification. So perhaps OpenNeuro would look like:

```
bids://openneuro.org/ds<accession>.v<version>:/<relpath>
bids://openneuro.org/ds000003.v2.0.1:/sub-01/anat/sub-01_desc-preproc_T1w.nii.gz
```

Note that the `/` after the authority is needed to ensure that the dataset name is part of the path, not part of the authority:

```Python
>>> from urllib import urlparse
>>> urlparse("bids://openneuro.org:ds000003.v2.0.1:/sub-01/anat/sub-01_desc-preproc_T1w.nii.gz")
ParseResult(scheme='bids', netloc='openneuro.org:ds000003.v2.0.1:', path='/sub-01/anat/sub-01_desc-preproc_T1w.nii.gz', params='', query='', fragment='')
>>> urlparse("bids://openneuro.org/ds000003.v2.0.1:/sub-01/anat/sub-01_desc-preproc_T1w.nii.gz")
ParseResult(scheme='bids', netloc='openneuro.org', path='/ds000003.v2.0.1:/sub-01/anat/sub-01_desc-preproc_T1w.nii.gz', params='', query='', fragment='')
```

Note that this doesn't remove the problem of resolution, but pushes it onto an authority. We would need a minimal protocol to at least say how to make a request, and I don't think we're prepared to write that text. I'm not, anyway.

I would suggest that we stick with the local URI mapping for now. It should be forwards-compatible with adding authorities. If this is something where a database is motivated to provide a reference implementation, it would not be difficult to write a small BEP as a provisional spec, wait on the implementation(s), and then merge it fully-baked.